File: auth-service/handlers.go
package main

import (
	"encoding/json"
	"log"
	"net/http"
	"strings"

	"golang.org/x/crypto/bcrypt"
)

// LoginRequest represents the JSON payload for login
type LoginRequest struct {
	Username string `json:"username"`
	Password string `json:"password"`
}

// LoginResponse represents the successful login response
type LoginResponse struct {
	Token string `json:"token"`
}

// ErrorResponse represents an error response
type ErrorResponse struct {
	Error string `json:"error"`
	Code  string `json:"code"`
}

// handleLogin processes POST /login requests
// This endpoint performs CPU-intensive bcrypt comparison and JWT generation
func handleLogin(w http.ResponseWriter, r *http.Request) {
	// Only accept POST requests
	if r.Method != http.MethodPost {
		w.WriteHeader(http.StatusMethodNotAllowed)
		json.NewEncoder(w).Encode(ErrorResponse{
			Error: "Method not allowed",
			Code:  "METHOD_NOT_ALLOWED",
		})
		return
	}

	// Parse JSON request body
	var req LoginRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		log.Printf("[AUTH] ERROR: Invalid JSON in login request: %v", err)
		w.WriteHeader(http.StatusBadRequest)
		json.NewEncoder(w).Encode(ErrorResponse{
			Error: "Invalid request format",
			Code:  "INVALID_JSON",
		})
		return
	}

	log.Printf("[AUTH] Login attempt for user: %s", req.Username)

	// Look up user in database (I/O operation)
	user, err := getUserByUsername(req.Username)
	if err != nil {
		log.Printf("[AUTH] ERROR: Database error during login: %v", err)
		w.WriteHeader(http.StatusInternalServerError)
		json.NewEncoder(w).Encode(ErrorResponse{
			Error: "Internal server error",
			Code:  "DATABASE_ERROR",
		})
		return
	}

	if user == nil {
		log.Printf("[AUTH] FAILED: User not found: %s", req.Username)
		w.WriteHeader(http.StatusUnauthorized)
		json.NewEncoder(w).Encode(ErrorResponse{
			Error: "Invalid credentials",
			Code:  "INVALID_CREDENTIALS",
		})
		return
	}

	// ========================================
	// CPU-INTENSIVE OPERATION: bcrypt comparison
	// This is why we isolated this service from Node.js
	// ========================================
	log.Printf("[AUTH] Performing bcrypt comparison (CPU-intensive)...")
	err = bcrypt.CompareHashAndPassword([]byte(user.PasswordHash), []byte(req.Password))
	if err != nil {
		log.Printf("[AUTH] FAILED: Invalid password for user: %s", req.Username)
		w.WriteHeader(http.StatusUnauthorized)
		json.NewEncoder(w).Encode(ErrorResponse{
			Error: "Invalid credentials",
			Code:  "INVALID_CREDENTIALS",
		})
		return
	}

	// ========================================
	// CPU-INTENSIVE OPERATION: JWT generation
	// HMAC-SHA256 signing is computationally expensive
	// ========================================
	log.Printf("[AUTH] Generating JWT token (CPU-intensive)...")
	token, err := generateJWT(user.ID)
	if err != nil {
		log.Printf("[AUTH] ERROR: Failed to generate JWT: %v", err)
		w.WriteHeader(http.StatusInternalServerError)
		json.NewEncoder(w).Encode(ErrorResponse{
			Error: "Internal server error",
			Code:  "TOKEN_GENERATION_FAILED",
		})
		return
	}

	log.Printf("[AUTH] SUCCESS: User %s authenticated, JWT generated", req.Username)

	// Return JWT token to client
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusOK)
	json.NewEncoder(w).Encode(LoginResponse{
		Token: token,
	})
}

// handleValidate processes GET /validate requests from Nginx
// This endpoint is called via auth_request to validate JWT tokens
func handleValidate(w http.ResponseWriter, r *http.Request) {
	// Only accept GET requests
	if r.Method != http.MethodGet {
		w.WriteHeader(http.StatusMethodNotAllowed)
		return
	}

	// Extract Authorization header
	authHeader := r.Header.Get("Authorization")
	if authHeader == "" {
		log.Println("[AUTH] VALIDATION FAILED: Missing Authorization header")
		w.WriteHeader(http.StatusUnauthorized)
		return
	}

	// Check for Bearer token format
	parts := strings.Split(authHeader, " ")
	if len(parts) != 2 || parts[0] != "Bearer" {
		log.Println("[AUTH] VALIDATION FAILED: Invalid Authorization header format")
		w.WriteHeader(http.StatusUnauthorized)
		return
	}

	tokenString := parts[1]

	// ========================================
	// CPU-INTENSIVE OPERATION: JWT validation
	// HMAC-SHA256 signature verification is computationally expensive
	// ========================================
	log.Println("[AUTH] Validating JWT token (CPU-intensive)...")
	userID, err := validateJWT(tokenString)
	if err != nil {
		log.Printf("[AUTH] VALIDATION FAILED: %v", err)
		w.WriteHeader(http.StatusUnauthorized)
		return
	}

	log.Printf("[AUTH] VALIDATION SUCCESS: Token valid for user ID %d", userID)

	// CRITICAL: Inject X-User-ID header for Nginx to forward to API service
	// This is the trusted header that the Node.js service will read
	w.Header().Set("X-User-ID", extractUserIDString(userID))
	w.WriteHeader(http.StatusOK)
}

File: auth-service-rust/src/handlers.rs
use axum::{
    extract::State,
    http::{HeaderMap, StatusCode},
    Json,
};
use bcrypt::verify;
use sqlx::PgPool;

use crate::models::{ErrorResponse, LoginRequest, LoginResponse};
use crate::{db, jwt};

/// Application state containing shared resources
#[derive(Clone)]
pub struct AppState {
    pub db_pool: PgPool,
    pub jwt_secret: String,
}

/// Handle POST /login requests
/// Authenticates user and returns JWT token
pub async fn handle_login(
    State(state): State<AppState>,
    Json(req): Json<LoginRequest>,
) -> Result<Json<LoginResponse>, (StatusCode, Json<ErrorResponse>)> {
    println!("[AUTH] Login attempt for user: {}", req.username);

    // Look up user in database (I/O operation)
    let user = match db::get_user_by_username(&state.db_pool, &req.username).await {
        Ok(Some(user)) => user,
        Ok(None) => {
            println!("[AUTH] FAILED: User not found: {}", req.username);
            return Err((
                StatusCode::UNAUTHORIZED,
                Json(ErrorResponse {
                    error: "Invalid credentials".to_string(),
                    code: "INVALID_CREDENTIALS".to_string(),
                }),
            ));
        }
        Err(e) => {
            println!("[AUTH] ERROR: Database error during login: {}", e);
            return Err((
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(ErrorResponse {
                    error: "Internal server error".to_string(),
                    code: "DATABASE_ERROR".to_string(),
                }),
            ));
        }
    };

    // ========================================
    // CPU-INTENSIVE OPERATION: bcrypt comparison
    // This is why we isolated this service from Node.js
    // ========================================
    let password_match = match verify(&req.password, &user.password_hash) {
        Ok(is_match) => is_match,
        Err(e) => {
            println!("[AUTH] ERROR: bcrypt verification failed: {}", e);
            return Err((
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(ErrorResponse {
                    error: "Internal server error".to_string(),
                    code: "HASH_ERROR".to_string(),
                }),
            ));
        }
    };

    if !password_match {
        println!("[AUTH] FAILED: Invalid password for user: {}", req.username);
        return Err((
            StatusCode::UNAUTHORIZED,
            Json(ErrorResponse {
                error: "Invalid credentials".to_string(),
                code: "INVALID_CREDENTIALS".to_string(),
            }),
        ));
    }

    // ========================================
    // CPU-INTENSIVE OPERATION: JWT signing
    // ========================================
    let token = match jwt::generate_jwt(user.id, &state.jwt_secret) {
        Ok(token) => token,
        Err(e) => {
            println!("[AUTH] ERROR: JWT generation failed: {}", e);
            return Err((
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(ErrorResponse {
                    error: "Internal server error".to_string(),
                    code: "JWT_ERROR".to_string(),
                }),
            ));
        }
    };

    println!("[AUTH] SUCCESS: User {} authenticated", req.username);
    Ok(Json(LoginResponse { token }))
}

/// Handle GET /validate requests
/// Validates JWT token (called by Nginx via auth_request directive)
/// Returns 200 if valid, otherwise returns error status
pub async fn handle_validate(
    State(state): State<AppState>,
    headers: HeaderMap,
) -> Result<StatusCode, (StatusCode, Json<ErrorResponse>)> {
    // Extract Authorization header
    let auth_header = match headers.get("Authorization") {
        Some(value) => match value.to_str() {
            Ok(header) => header,
            Err(_) => {
                println!("[AUTH] VALIDATION FAILED: Invalid Authorization header encoding");
                return Err((
                    StatusCode::UNAUTHORIZED,
                    Json(ErrorResponse {
                        error: "Invalid Authorization header".to_string(),
                        code: "INVALID_HEADER".to_string(),
                    }),
                ));
            }
        },
        None => {
            println!("[AUTH] VALIDATION FAILED: Missing Authorization header");
            return Err((
                StatusCode::UNAUTHORIZED,
                Json(ErrorResponse {
                    error: "Missing Authorization header".to_string(),
                    code: "MISSING_TOKEN".to_string(),
                }),
            ));
        }
    };

    // Parse "Bearer <token>" format
    let token = match auth_header.strip_prefix("Bearer ") {
        Some(t) => t,
        None => {
            println!("[AUTH] VALIDATION FAILED: Invalid Authorization header format");
            return Err((
                StatusCode::UNAUTHORIZED,
                Json(ErrorResponse {
                    error: "Invalid Authorization header format".to_string(),
                    code: "INVALID_FORMAT".to_string(),
                }),
            ));
        }
    };

    println!("[AUTH] Validating token...");

    // ========================================
    // CPU-INTENSIVE OPERATION: JWT verification
    // Validates signature and expiration
    // ========================================

    match jwt::validate_jwt(token, &state.jwt_secret) {
        Ok(claims) => {
            println!(
                "[AUTH] SUCCESS: Token validated for user {}",
                claims.user_id()
            );
            claims
        }
        Err(e) => {
            println!("[AUTH] FAILED: Token validation error: {}", e);
            return Err((
                StatusCode::UNAUTHORIZED,
                Json(ErrorResponse {
                    error: "Invalid token".to_string(),
                    code: "INVALID_TOKEN".to_string(),
                }),
            ));
        }
    };

    // Return 200 OK - JWT validation successful
    // Nginx will allow the request to proceed. The API service will validate the JWT claims.
    Ok(StatusCode::OK)
}
